lib/Tools/LinearLayout.cpp

```diff
@@ -804,118 +780,179 @@ LinearLayout LinearLayout::compose(const LinearLayout &outer) const {
+      basis |= (retMat[r] >> c & 1) << r;
     }
     bs.push_back({basis});
   }
 
-  LinearLayout flatComposed(std::move(newBases),
-                            {{outDim1D, outer.getTotalInDimSize()}},
+  LinearLayout retFlattened(std::move(retBases),
+                            {{outDim1D, A.getTotalInDimSize()}},
                             /*requireSurjective=*/false);
 
   SmallVector<std::pair<StringAttr, int32_t>> retInDims;
   SmallVector<std::pair<StringAttr, int32_t>> retOutDims;
-  for (StringAttr dim : getInDimNames()) {
-    retInDims.push_back({dim, getInDimSize(dim)});
+  for (StringAttr dim : B.getInDimNames()) {
+    retInDims.push_back({dim, B.getInDimSize(dim)});
   }
-  for (StringAttr dim : outer.getInDimNames()) {
-    retOutDims.push_back({dim, outer.getInDimSize(dim)});
+  for (StringAttr dim : A.getInDimNames()) {
+    retOutDims.push_back({dim, A.getInDimSize(dim)});
   }
-  return flatComposed.reshapeIns(retInDims).reshapeOuts(retOutDims);
+  return retFlattened.reshapeIns(retInDims).reshapeOuts(retOutDims);
+}
+
+} // namespace
+
+LinearLayout LinearLayout::invertAndCompose(const LinearLayout &outer) const {
+  // TODO(Lezcano) Make friend and perhaps rename to `convertFrom`
+  // For this, we need to implement our LLVM lowerings by inverting the "from"
+  // layout, and then iterating over the elements from the "to" layout and
+  // fetching the corresponding element from the "from" layout. This exercises
+  // the broadcasting that we incentivise via choosing the minimum norm solution
+  // in lstsq.
+
+  // The order of dims does not matter. We choose to transpose outer
+  auto outDims = llvm::to_vector(getOutDimNames());
+  assertDimsEqualIgnoringOrder(outDims, outer.getOutDimNames());
+  const auto &B = *this;
+  const auto A = outer.transposeOuts(outDims);
+  for (auto dim : outDims) {
+    assert(A.getOutDimSize(dim) == B.getOutDimSize(dim) &&
+           "Convert layout does not change the shape of a tensor");
+  }
+
+  // We'll write A^{-1} to mean the inverse or the pseudo-inverse of A
+  // We are computing A^{-1}B so from must be surjective so that
+  // it has a left inverse.
+  assert(A.isSurjective());
+
+  // Broadcasting heuristic
+  // Imagine we have two layouts with `warps = [[0, 0],  [0, 0]]`
+  // (broadcasting) on both layouts. We could map any warp to any warp in the
+  // conversion. Now, we want to map them as the identity map, to mark that
+  // nothing needs to be done there (`lstsq` would map all the warps to the
+  // zero warp, minimum norm solution). The heuristic here is as follows:
+  // - If a dimension is the same for both layouts, we want to map it as the
+  // identity
+  //   Equivalently, we don't add it to the conversion
+  // - Otherwise, we just call lstsq (i.e. map all the equivalent elements
+  //   to the same input element) to take advantage of broadcasting in shared
+  //   memory and avoid saving repeated elements in shared memory
```

Comment:
What does it mean broadcasting in “shared memory", do we really duplicate values in shared memory?

Reply:
I meant the broadcasting in the reads, as we do now. But more generally, this gives us the following behaviour:

Before, going to shared memory, we always did `shmem^{-1} · distributed` when going to shared memory, we crafted `shmem` to be invertible, so the inverse wors just fine, and then we use two facts:
- That when going to shmem, if multiple threads write to the same location there'll be a race condition and a bank conflict, but this is logically fine because we know that they all have the same data. This race condition is not great perf-wise as they need to be serialised tho.
- When reading from shmem, broadcasting happens just as expected.

In the future, we will always want to form the convert layout that goes from the `to` layout to the `from` layout. This means that we will iterate the elements of the output and we ask "which element from the input do I need to pull this from?". This is how we do the reads from shmem, but for the writes we'll need to change the code, but having the zeros above will give us an approach that will take most advantage of predicated stores that will avoid this stalls, which is nice.

Reply:
This happens when there is a read, but aren't writes to the same memory address serialised?

Reply:
At least it's not called a bank conflict as far as I know. ncu won't report any bank conflicts if all writes go to the same address